File: core/image/hsb_mapper.py
"""
HSB 映射器 v2.3.2
游戏内选色器非线性感知曲线的校准模型。
全空间暴力建表 (201x213x112) + 欧氏距离最近邻搜索。
LUT 缓存至 assets/hsb_lut.npz，首次生成后续复用。
"""

import os
import numpy as np

from core.utils.logger import get_logger
from core.utils.resource import resource_path

logger = get_logger("hsb_mapper")

HUE_ANCHORS = [
    (0,   "#FF0000"),
    (34,  "#FF00FF"),
    (64,  "#0000FF"),
    (100, "#00FFFF"),
    (136, "#00FF00"),
    (166, "#FFFF00"),
    (200, "#FF0000"),
]

H_MAX = 200
S_MAX = 212
V_MAX = 111

LUT_PATH = "assets/hsb_lut.npz"


def _hex_to_rgb01(hex_str: str) -> np.ndarray:
    h = hex_str.lstrip("#")
    return np.array([int(h[i:i+2], 16) / 255.0 for i in (0, 2, 4)], dtype=np.float64)


def _rgb01_to_h(rgb: np.ndarray) -> float:
    r, g, b = rgb
    mn = min(r, g, b)
    mx = max(r, g, b)
    d = mx - mn
    if d == 0:
        return 0.0
    if mx == r:
        return 60.0 * (((g - b) / d) % 6)
    if mx == g:
        return 60.0 * ((b - r) / d + 2)
    return 60.0 * ((r - g) / d + 4)


def _hsv_to_rgb01(h, s, v):
    if s == 0:
        return np.array([v, v, v], dtype=np.float64)
    h = h / 60.0
    i = int(h)
    f = h - i
    p = v * (1 - s)
    q = v * (1 - s * f)
    t = v * (1 - s * (1 - f))
    lut = {
        0: (v, t, p), 1: (q, v, p), 2: (p, v, t),
        3: (p, q, v), 4: (t, p, v), 5: (v, p, q),
    }
    return np.array(lut.get(i % 6, (0, 0, 0)), dtype=np.float64)


def _hue_angle_for_step(h_step):
    for i in range(len(HUE_ANCHORS) - 1):
        s0, h0 = HUE_ANCHORS[i]
        s1, h1 = HUE_ANCHORS[i + 1]
        if s0 <= h_step <= s1:
            if s1 == s0:
                break
            t = (h_step - s0) / (s1 - s0)
            hue0 = _rgb01_to_h(_hex_to_rgb01(h0))
            hue1 = _rgb01_to_h(_hex_to_rgb01(h1))
            if abs(hue1 - hue0) > 180:
                if hue0 < hue1:
                    hue0 += 360
                else:
                    hue1 += 360
            hue = hue0 + t * (hue1 - hue0)
            return hue % 360
    return 0.0


def _saturation_for_step(s_step):
    p = max(0.0, min(1.0, s_step / S_MAX))
    if p == 0:
        return 0.0
    s = 0.49 * p + 0.2 * (p ** 38) + 0.31 * (p ** 3.9)
    return max(0.0, min(1.0, s))


def _brightness_for_step(b_step):
    p = max(0.0, min(1.0, b_step / V_MAX))
    if p == 0:
        return 0.0
    return p ** (1.0 / 2.26)


def _build_lut():
    logger.info("构建 HSB LUT ...")
    lut = np.zeros((H_MAX + 1, S_MAX + 1, V_MAX + 1, 3), dtype=np.int16)
    for h_s in range(H_MAX + 1):
        hue = _hue_angle_for_step(h_s)
        for s_s in range(S_MAX + 1):
            sat = _saturation_for_step(s_s)
            for b_s in range(V_MAX + 1):
                val = _brightness_for_step(b_s)
                rgb = _hsv_to_rgb01(hue, sat, val)
                lut[h_s, s_s, b_s] = [int(round(rgb[i] * 255)) for i in range(3)]
        if h_s % 20 == 0:
            logger.info(f"  H: {h_s}/{H_MAX}")
    logger.info("HSB LUT 完成")
    return lut


_lut = None


def get_lut():
    global _lut
    if _lut is not None:
        return _lut
    rp = resource_path(LUT_PATH)
    if os.path.exists(rp):
        logger.info(f"加载 LUT: {rp}")
        _lut = np.load(rp)["lut"]
    else:
        _lut = _build_lut()
        os.makedirs(os.path.dirname(resource_path(LUT_PATH)), exist_ok=True)
        np.savez_compressed(resource_path(LUT_PATH), lut=_lut)
        logger.info(f"LUT 已保存")
    return _lut


def rgb_to_steps(r, g, b):
    lut = get_lut()
    diff = lut.astype(np.int32) - np.array([r, g, b], dtype=np.int32)
    dist_sq = diff[:, :, :, 0]**2 + diff[:, :, :, 1]**2 + diff[:, :, :, 2]**2
    idx = np.argmin(dist_sq)
    h_s, s_s, b_s = np.unravel_index(idx, (H_MAX + 1, S_MAX + 1, V_MAX + 1))
    return int(h_s), int(s_s), int(b_s)

File: frontend/src/engine/hsb-mapper.ts
/**
 * HSB 映射器 — 游戏内自定义选色器非线性校准。
 * 将 RGB 颜色映射为 ZR/右/上 按键次数 (h/s/b steps)。
 * 浏览器端纯 TypeScript 实现，使用多分辨率搜索替代 LUT。
 */

// 色相锚定点: (step, hex_color)
const HUE_ANCHORS: [number, string][] = [
  [0, '#FF0000'], [34, '#FF00FF'], [64, '#0000FF'],
  [100, '#00FFFF'], [136, '#00FF00'], [166, '#FFFF00'],
  [200, '#FF0000'],
];

export const H_MAX = 200;
export const S_MAX = 212;
export const V_MAX = 111;

// ---- 缓存 ----
const _reverseCache = new Map<number, [number, number, number]>();

function cacheKey(r: number, g: number, b: number): number {
  return (r << 16) | (g << 8) | b;
}

// ---- 正向映射: (h,s,b) steps → RGB ----

function hexToRgb01(hex: string): [number, number, number] {
  const h = hex.replace('#', '');
  return [
    parseInt(h.substring(0, 2), 16) / 255,
    parseInt(h.substring(2, 4), 16) / 255,
    parseInt(h.substring(4, 6), 16) / 255,
  ];
}

function rgb01ToH(r: number, g: number, b: number): number {
  const mn = Math.min(r, g, b);
  const mx = Math.max(r, g, b);
  const d = mx - mn;
  if (d === 0) return 0;
  if (mx === r) return 60 * (((g - b) / d) % 6);
  if (mx === g) return 60 * ((b - r) / d + 2);
  return 60 * ((r - g) / d + 4);
}

function hsvToRgb01(h: number, s: number, v: number): [number, number, number] {
  if (s === 0) return [v, v, v];
  const hh = h / 60;
  const i = Math.floor(hh);
  const f = hh - i;
  const p = v * (1 - s);
  const q = v * (1 - s * f);
  const t = v * (1 - s * (1 - f));
  const lut: [number, number, number][] = [
    [v, t, p], [q, v, p], [p, v, t],
    [p, q, v], [t, p, v], [v, p, q],
  ];
  return lut[i % 6] || [0, 0, 0];
}

function hueAngleForStep(hStep: number): number {
  for (let i = 0; i < HUE_ANCHORS.length - 1; i++) {
    const [s0, h0] = HUE_ANCHORS[i];
    const [s1, h1] = HUE_ANCHORS[i + 1];
    if (s0 <= hStep && hStep <= s1) {
      if (s1 === s0) break;
      const t = (hStep - s0) / (s1 - s0);
      let hue0 = rgb01ToH(...hexToRgb01(h0));
      let hue1 = rgb01ToH(...hexToRgb01(h1));
      if (Math.abs(hue1 - hue0) > 180) {
        if (hue0 < hue1) hue0 += 360;
        else hue1 += 360;
      }
      return (hue0 + t * (hue1 - hue0)) % 360;
    }
  }
  return 0;
}

function saturationForStep(sStep: number): number {
  const p = Math.max(0, Math.min(1, sStep / S_MAX));
  if (p === 0) return 0;
  return Math.max(0, Math.min(1, 0.49 * p + 0.2 * Math.pow(p, 38) + 0.31 * Math.pow(p, 3.9)));
}

function brightnessForStep(bStep: number): number {
  const p = Math.max(0, Math.min(1, bStep / V_MAX));
  if (p === 0) return 0;
  return Math.pow(p, 1.0 / 2.26);
}

/** 正向映射: (h,s,v) 步数值 → RGB [0-255] */
export function hsvStepsToRgb(hStep: number, sStep: number, bStep: number): [number, number, number] {
  const hue = hueAngleForStep(hStep);
  const sat = saturationForStep(sStep);
  const val = brightnessForStep(bStep);
  const [r01, g01, b01] = hsvToRgb01(hue, sat, val);
  return [
    Math.round(r01 * 255),
    Math.round(g01 * 255),
    Math.round(b01 * 255),
  ];
}

// ---- 反向映射: RGB → (h,s,b) steps ----

function sqDist(r1: number, g1: number, b1: number, r2: number, g2: number, b2: number): number {
  const dr = r1 - r2;
  const dg = g1 - g2;
  const db = b1 - b2;
  return dr * dr + dg * dg + db * db;
}

/**
 * 将 RGB 颜色映射为 (h_step, s_step, b_step)。
 * 使用多分辨率搜索（16步粗搜 + 精细搜索），结果缓存。
 */
export function rgbToSteps(r: number, g: number, b: number): [number, number, number] {
  const key = cacheKey(r, g, b);
  const cached = _reverseCache.get(key);
  if (cached) return cached;

  let bestH = 0, bestS = 0, bestV = 0;
  let bestDist = Infinity;

  // 粗搜：步长 16
  const COARSE = 16;
  for (let h = 0; h <= H_MAX; h += COARSE) {
    for (let s = 0; s <= S_MAX; s += COARSE) {
      for (let v = 0; v <= V_MAX; v += COARSE) {
        const [pr, pg, pb] = hsvStepsToRgb(h, s, v);
        const d = sqDist(r, g, b, pr, pg, pb);
        if (d < bestDist) { bestDist = d; bestH = h; bestS = s; bestV = v; }
      }
    }
  }

  // 精细搜索：粗搜最佳点周围 ±16
  for (let h = Math.max(0, bestH - COARSE); h <= Math.min(H_MAX, bestH + COARSE); h++) {
    for (let s = Math.max(0, bestS - COARSE); s <= Math.min(S_MAX, bestS + COARSE); s++) {
      for (let v = Math.max(0, bestV - COARSE); v <= Math.min(V_MAX, bestV + COARSE); v++) {
        const [pr, pg, pb] = hsvStepsToRgb(h, s, v);
        const d = sqDist(r, g, b, pr, pg, pb);
        if (d < bestDist) { bestDist = d; bestH = h; bestS = s; bestV = v; }
      }
    }
  }

  const result: [number, number, number] = [bestH, bestS, bestV];
  _reverseCache.set(key, result);
  return result;
}

/**
 * 批量 RGB → steps 映射（使用 Web Worker 避免阻塞 UI）。
 * 简单场景无需 worker，直接同步调用。
 */
export function rgbToStepsBatch(colors: [number, number, number][]): [number, number, number][] {
  return colors.map(([r, g, b]) => rgbToSteps(r, g, b));
}
